File: tests-playwright/pages/cart-page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base-page';

export class CartPage extends BasePage {
  private readonly cartContainer: Locator;
  private readonly cartItems: Locator;
  private readonly checkoutButton: Locator;
  private readonly continueShoppingButton: Locator;
  private readonly removeButtons: Locator;

  constructor(page: Page) {
    super(page);
    this.cartContainer = page.locator('#cart_contents_container');
    this.cartItems = page.locator('.cart_item');
    this.checkoutButton = page.locator('[data-test="checkout"]');
    this.continueShoppingButton = page.locator('[data-test="continue-shopping"]');
    this.removeButtons = page.locator('[data-test*="remove-"]');
  }

  async getCartItemsCount(): Promise<number> {
    return await this.cartItems.count();
  }

  async getItemTitle(index: number): Promise<string> {
    const item = this.cartItems.nth(index);
    const titleElement = item.locator('.inventory_item_name');
    return await this.getElementText(titleElement);
  }

  async getItemPrice(index: number): Promise<string> {
    const item = this.cartItems.nth(index);
    const priceElement = item.locator('.inventory_item_price');
    return await this.getElementText(priceElement);
  }

  async getItemQuantity(index: number): Promise<string> {
    const item = this.cartItems.nth(index);
    const quantityElement = item.locator('.cart_quantity');
    return await this.getElementText(quantityElement);
  }

  async removeItem(index: number): Promise<void> {
    const removeButton = this.removeButtons.nth(index);
    await this.clickElement(removeButton);
  }

  async proceedToCheckout(): Promise<void> {
    await this.clickElement(this.checkoutButton);
  }

  async continueShopping(): Promise<void> {
    await this.clickElement(this.continueShoppingButton);
  }

  async isCartPageLoaded(): Promise<boolean> {
    return await this.isElementVisible(this.cartContainer);
  }

  async isCartEmpty(): Promise<boolean> {
    return await this.getCartItemsCount() === 0;
  }
}

File: tests-csharp/Pages/CartPage.cs
using Microsoft.Playwright;
using Tests.CSharp.Pages;

namespace Tests.CSharp.Pages;

public class CartPage : BasePage
{
    private readonly ILocator _cartContainer;
    private readonly ILocator _cartItems;
    private readonly ILocator _checkoutButton;
    private readonly ILocator _continueShoppingButton;
    private readonly ILocator _removeButtons;

    public CartPage(IPage page) : base(page)
    {
        _cartContainer = Page.Locator("#cart_contents_container");
        _cartItems = Page.Locator(".cart_item");
        _checkoutButton = Page.Locator("[data-test='checkout']");
        _continueShoppingButton = Page.Locator("[data-test='continue-shopping']");
        _removeButtons = Page.Locator("[data-test*='remove-']");
    }

    public async Task<int> GetCartItemsCountAsync()
    {
        return await _cartItems.CountAsync();
    }

    public async Task<string> GetItemTitleAsync(int index)
    {
        var item = _cartItems.Nth(index);
        var titleElement = item.Locator(".inventory_item_name");
        return await GetElementTextAsync(titleElement);
    }

    public async Task<string> GetItemPriceAsync(int index)
    {
        var item = _cartItems.Nth(index);
        var priceElement = item.Locator(".inventory_item_price");
        return await GetElementTextAsync(priceElement);
    }

    public async Task<string> GetItemQuantityAsync(int index)
    {
        var item = _cartItems.Nth(index);
        var quantityElement = item.Locator(".cart_quantity");
        return await GetElementTextAsync(quantityElement);
    }

    public async Task RemoveItemAsync(int index)
    {
        var removeButton = _removeButtons.Nth(index);
        await ClickElementAsync(removeButton);
    }

    public async Task ProceedToCheckoutAsync()
    {
        await ClickElementAsync(_checkoutButton);
    }

    public async Task ContinueShoppingAsync()
    {
        await ClickElementAsync(_continueShoppingButton);
    }

    public async Task<bool> IsCartPageLoadedAsync()
    {
        return await IsElementVisibleAsync(_cartContainer);
    }

    public async Task<bool> IsCartEmptyAsync()
    {
        return await GetCartItemsCountAsync() == 0;
    }
}

